core/engine/src/object/jsobject.rs

```diff
@@ -24,38 +24,64 @@ use crate::{
     property::{PropertyDescriptor, PropertyKey},
     value::PreferredType,
 };
-use boa_gc::{self, Finalize, Gc, GcBox, GcRefCell, Trace};
+use boa_gc::{self, Finalize, Gc, GcErased, GcRefCell, Trace};
 use std::collections::HashSet;
 use std::{
     cell::RefCell,
     collections::HashMap,
     error::Error,
     fmt::{self, Debug, Display},
     hash::Hash,
-    ptr::NonNull,
     result::Result as StdResult,
 };
 use thin_vec::ThinVec;
 
+#[cfg(not(feature = "jsvalue-enum"))]
+use boa_gc::GcBox;
+
+#[cfg(not(feature = "jsvalue-enum"))]
+use std::ptr::NonNull;
+
 /// A wrapper type for an immutably borrowed type T.
 pub type Ref<'a, T> = boa_gc::GcRef<'a, T>;
 
 /// A wrapper type for a mutably borrowed type T.
 pub type RefMut<'a, T, U> = boa_gc::GcRefMut<'a, T, U>;
 
-/// An `Object` with inner data set to `dyn NativeObject`.
-pub type ErasedObject = Object<dyn NativeObject>;
+pub(crate) type ErasedVTableObject = VTableObject<ErasedObjectData>;
+
+/// An `Object` with inner data set to `ErasedObjectData`.
+pub type ErasedObject = Object<ErasedObjectData>;
 
-pub(crate) type ErasedVTableObject = VTableObject<dyn NativeObject>;
+/// A erased object data type that must never be used directly.
+#[derive(Debug, Trace, Finalize)]
+pub struct ErasedObjectData {}
+
+impl JsData for ErasedObjectData {}
 
 /// Garbage collected `Object`.
 #[derive(Trace, Finalize)]
 #[boa_gc(unsafe_no_drop)]
-pub struct JsObject<T: NativeObject + ?Sized = dyn NativeObject> {
-    inner: Gc<VTableObject<T>>,
+pub struct JsObject {
+    inner: GcErased,
+}
+
+impl Clone for JsObject {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+/// Garbage collected typed `Object`.
+#[derive(Trace, Finalize)]
+#[boa_gc(unsafe_no_drop)]
+pub struct JsObjectTyped<T: NativeObject> {
+    pub(crate) inner: Gc<VTableObject<T>>,
 }
```

Comment:
```suggestion
#[derive(Trace, Finalize)]
#[boa_gc(unsafe_no_drop)]
pub struct JsObjectTyped<T: NativeObject = ErasedObjectData> {
    pub(crate) inner: Gc<VTableObject<T>>,
}
```

Couldn't we have a default parameter here and it would eliminate the need for two `JsObject` types.

Reply:
Yes, we could keep it like this and avoid `JsObjectTyped`. I just wanted to try it out like this to leverage `GcErased`. As per our discussion, I personally don't think there is a big benefit to splitting the erased and typed variants, so unless someone can think of a big benefit, I would also be for keeping the default paramter.

Reply:
> We could still use GcErased in JsObject (if we wanted). It would just requires having a small method that we used to access inner as VTableObject<T> (by casting with GcErased::cast).

And we would need a `PhantomData<T>`, right? That would be my reason just to use `Gc<T>` directly, since the code would be easier.

Reply:
> And we would need a `PhantomData<T>`, right? That would be my reason just to use `Gc<T>` directly, since the code would be easier.

Yes, phantom data is required to appease the compiler from the warning of unused `T`. Yes, using `Gc<T>` will be simpler :smile: